src/otaclient_common/retry_task_map.py

```diff
@@ -192,7 +215,20 @@ def _fut_gen(self, interval: float) -> Generator[Future[Any]]:
                 with contextlib.suppress(Empty):
                     while True:  # drain the _fut_queue
                         self._fut_queue.get_nowait()
-                raise TasksEnsureFailed(self._failure_msg)  # raise exc to upper caller
+
+                try:
+                    _wrapped_exc = self._exc_deque.pop()
+                except Exception:
+                    _wrapped_exc = TasksEnsureFailed(
+                        "execution interrupted due to thread pool shutdown"
+                    )
+
+                # NOTE: still raise TaskEnsureFailed to upper,
+                #       let choose to upper dig into the TaskEnsureFailed or not.
+                try:  # raise exc to upper caller
+                    raise _wrapped_exc
+                finally:
+                    del _wrapped_exc
```

Comment:
do we need to del this local variable clearly? I think `raise _wrapped_exc` will be enough.

Reply:
To be noticed that, the `try ... except: raise` statement has special treatment to the above condition, python will break the circular reference for us.

Reply:
Thank you!! very deep and useful knowledge :+1: